File: nprs_app/android/NumberPlateRecognition/src/main/java/com/nprs/app/tools/ImageConverter.java
//package com.nprs.app.tools;
//
//import android.graphics.Bitmap;
//import com.nprs.recsystem.common.exceptions.ArgumentException;
//import com.nprs.recsystem.common.image.ColorFormat;
//import com.nprs.recsystem.common.image.ColorInfo;
//import com.nprs.recsystem.common.image.Image;
//
//
//public class ImageConverter {
//    public static Image bmpToRsImage(Bitmap bmp) {
//        Image res = new Image(bmp.getWidth(), bmp.getHeight(), ColorInfo.Rgb);
//        int[] pixels = new int[bmp.getWidth() * bmp.getHeight()];
//        bmp.getPixels(pixels, 0, bmp.getWidth(), 0, 0, bmp.getWidth(), bmp.getHeight());
//
//        for(int x = 0; x < bmp.getWidth(); x++) {
//            for (int y = 0; y < bmp.getHeight(); y++) {
//                int c = pixels[x + bmp.getWidth() * y];
//                res.setValue(x, y, 0, android.graphics.Color.red(c));
//                res.setValue(x, y, 1, android.graphics.Color.green(c));
//                res.setValue(x, y, 2, android.graphics.Color.blue(c));
//            }
//        }
//
//        return res;
//    }
//
//    public static Bitmap rsImageToBitmap(Image image) throws ArgumentException {
//        int[] colors = new int[image.width() * image.height()];
//
//        for (int x = 0; x < image.width(); x++) {
//            for (int y = 0; y < image.height(); y++) {
//                try {
//                    colors[x + y * image.width()] = colorToInt(image, x, y);
//                } catch (ArgumentException e) {
//                    throw new ArgumentException("ImageConverter.rsImageToBitmap: Failed to convert image to bitmap. " + e.getMessage());
//                }
//            }
//        }
//
//        return Bitmap.createBitmap(colors, image.width(), image.height(), Bitmap.Config.ARGB_8888);
//    }
//
//    private static int colorToInt(Image image, int x, int y) throws ArgumentException {
//        int res = -1;
//
//        if (image.numChannels() == 3 && image.colorInfo().colorFormat() == ColorFormat.RGB) {
//            res = android.graphics.Color.argb(255, (int) (image.getValue(x, y, 0)),
//                    (int) (image.getValue(x, y, 1)),
//                    (int) (image.getValue(x, y, 2)));
//        }
//        else if (image.numChannels() == 1 && image.colorInfo().colorFormat() == ColorFormat.INTENSITY) {
//            int value = (int) (255 * image.getValue(x, y, 0));
//            res = android.graphics.Color.argb(255, value, value, value);
//        }
//        else {
//            throw new ArgumentException("ImageConverter.colorToInt(): Unknown image format");
//        }
//
//        return res;
//    }
//}

File: nprs_lib/sources/common/src/common/image/ImageConverter.cpp
#include "ImageConverter.h"
#include <common/exceptions/CommonExceptions.h>

using namespace nprs;

Image ImageConverter::convertRaw(const Bitmap &raw) {
    switch (raw.colorInfo().format()) {
    case ColorFormat::RGB: return rgbToInt(raw.data(), raw.width(), raw.height());
    case ColorFormat::BGRA: return bgraToInt(raw.data(), raw.width(), raw.height());
    case ColorFormat::RGBA: return rgbaToInt(raw.data(), raw.width(), raw.height());
    case ColorFormat::ARGB: return argb32ToInt(raw.data(), raw.width(), raw.height());
    default:
        throw ArgumentException("Color format not supported");
    }
}

Bitmap ImageConverter::imageToRawBgra(const Image &image) {
    Bitmap result(image.width(), image.height(), ColorInfo(ColorFormat::BGRA, 4));
    uint8_t *data = result.data();

    for (int y = 0; y < image.height(); y++) {
        for (int x = 0; x < image.width(); x++) {
            data[0] = image(x, y, 0) * 255;
            data[1] = image(x, y, 0) * 255;
            data[2] = image(x, y, 0) * 255;
            data[3] = 255;
            data += result.colorInfo().numChannels();
        }
    }
    return result;
}

Bitmap ImageConverter::imageToRawRgb(const Image &image) {
    Bitmap result(image.width(), image.height(), ColorInfo(ColorFormat::RGB, 3));
    uint8_t *data = result.data();

    for (int y = 0; y < image.height(); y++) {
        for (int x = 0; x < image.width(); x++) {
            data[0] = image(x, y, 0) * 255;
            data[1] = image(x, y, 0) * 255;
            data[2] = image(x, y, 0) * 255;
            data += result.colorInfo().numChannels();
        }
    }
    return result;
}

Bitmap ImageConverter::imageToRawArgb32(const Image &image) {
    Bitmap result(image.width(), image.height(), ColorInfo(ColorFormat::RGB, 3));
    uint8_t *data = result.data();

    for (int y = 0; y < image.height(); y++) {
        for (int x = 0; x < image.width(); x++) {
            data[0] = 255;
            data[1] = image(x, y, 0) * 255;
            data[2] = image(x, y, 0) * 255;
            data[3] = image(x, y, 0) * 255;
            data += result.colorInfo().numChannels();
        }
    }
    return result;
}


Image ImageConverter::bgraToInt(const uint8_t *data, int width, int height) {
    const uint8_t *data_as_byte = data;
    Image result(width, height, ColorInfo(ColorFormat::INTENSITY, 1));

    for (int y = 0; y < result.height(); y++) {
        for (int x = 0; x < result.width(); x++) {
            uint8_t r = data[2];
            uint8_t g = data[1];
            uint8_t b = data[0];
            data_as_byte += 4;
            result(x, y, 0) = 0.2126 * r / 255.0 + 0.7152 * g / 255.0 + 0.0722 * b / 255.0;
        }
    }
    return result;
}

Image ImageConverter::rgbToInt(const uint8_t *data, int width, int height) {
    const uint8_t *data_as_byte = data;
    Image result(width, height, ColorInfo(ColorFormat::INTENSITY, 1));

    for (int x = 0; x < result.width(); x++) {
        for (int y = 0; y < result.height(); y++) {
            uint8_t r = data[2];
            uint8_t g = data[1];
            uint8_t b = data[0];
            data_as_byte += 3;

            result(x, y, 0) = 0.2126 * r / 255.0 + 0.7152 * g / 255.0 + 0.0722 * b / 255.0;
        }
    }

    return result;
}

Image ImageConverter::rgbaToInt(const uint8_t *data, int width, int height) {
    Image result(width, height, ColorInfo(ColorFormat::INTENSITY, 1));
    const uint8_t *data_as_byte = data;

    for (int x = 0; x < result.width(); x++) {
        for (int y = 0; y < result.height(); y++) {
            uint8_t r = data[2];
            uint8_t g = data[1];
            uint8_t b = data[0];
            data_as_byte += 4;

            result(x, y, 0) = 0.2126 * r / 255.0 + 0.7152 * g / 255.0 + 0.0722 * b / 255.0;
        }
    }

    return result;
}

Image ImageConverter::argb32ToInt(uint8_t const* data, int width, int height) {
    const uint8_t *data_as_byte = data;
    Image result(width, height, ColorInfo(ColorFormat::INTENSITY, 1));

    for (int x = 0; x < result.width(); x++) {
        for (int y = 0; y < result.height(); y++) {
            uint8_t r = data[2];
            uint8_t g = data[1];
            uint8_t b = data[0];
            data_as_byte += 4;
            result(x, y, 0) = 0.2126 * r / 255.0 + 0.7152 * g / 255.0 + 0.0722 * b / 255.0;
        }
    }

    return result;
}